pkg/contrib/src/systems/NetcodeSystem.ts

```diff
@@ -0,0 +1,88 @@
+import { Entity, EntityProps, Game, GameProps, NetManager, System, SystemProps } from "@piggo-legends/core";
+import { Controller, Position } from "@piggo-legends/contrib";
+
+export type TickData = {
+  type: "game",
+  tick: number,
+  entities: Record<string, SerializedEntity>
+}
+
+export type SerializedEntity = {
+  position: {
+    x: number,
+    y: number
+  }
+}
+
+export type NetcodeSystemProps = SystemProps & {
+  net: NetManager
+}
+
+export class NetcodeSystem extends System<NetcodeSystemProps> {
+  componentTypeQuery = ["position"];
+
+  tickDataBuffer: TickData;
+
+  constructor(props: NetcodeSystemProps) {
+    super(props);
+    this.init();
+  }
+
+  init = () => {
+    this.props.net.events.addEventListener("message", (event: CustomEvent<any>) => {
+      if (event.detail.type === "game") this.tickDataBuffer = event.detail as TickData;
+    });
+  }
+
+  onTick = (entities: Entity<EntityProps>[], game: Game<GameProps>) => {
+    const serializedEntitites: Record<string, SerializedEntity> = {};
+
+    // serialize each entity
+    for (const entity of Object.values(entities)) {
+      const position = entity.components.position as Position;
+      const serialized: SerializedEntity = {
+        position: {
+          x: position.x, y: position.y
+        }
+      }
+      serializedEntitites[entity.id] = serialized;
+    }
+
+    // construct tick message
+    const message: TickData = {
+      type: "game",
+      tick: game.tick,
+      entities: serializedEntitites
+    }
+
+    if (game.tick === 100) console.log("sending", message);
+
+    // send message if connected
+    if (this.props.net.pc.connectionState === "connected") this.props.net.sendMessage(message);
+
+    // handle incoming message
+    if (this.tickDataBuffer) this.handleMessage(this.tickDataBuffer, game);
+  }
+
+  handleMessage = (td: TickData, game: Game<GameProps>) => {
+    if (td.tick % 400 === 0) console.log(td);
+
+    // Object.entries(td.entities).forEach(([id, entity]) => {
+    //   if (game.props.entities[id]) {
+    //     const position = game.props.entities[id].components.position as Position;
+    //     position.x = entity.position.x;
+    //     position.y = entity.position.y;
+    //   }
+    // });
+
+    const spaceship = game.props.entities["spaceship"] as Entity<EntityProps>;
+    const spaceshipController = spaceship.components.controller as Controller;
+
+    if (!spaceshipController.active) {
+      const position = spaceship.components.position as Position;
+
+      position.x = td.entities["spaceship"].position.x;
+      position.y = td.entities["spaceship"].position.y;
```

Comment:
how can this be extended to all networked entities? exclusive player control over any entity?